### ISSUE TITLE
lambda: corrupt zip archive asset produced when using node v16.13.1 and CDK v2.1.0

### ISSUE THREAD

**Issue by SamuelCabralCruz** [aws/aws-cdk#17976]
### What is the problem?

Resource handler returned message: "Uploaded file must be a non-empty zip (Service: Lambda, Status Code: 400, Request ID: 17f6356b-135a-4888-853a-f3a25b58fc70, Extended Request ID: null)" (RequestToken: e1d8746c-a6f9-a802-6bff-5de98eba863b, HandlerError
Code: InvalidRequest)

### Reproduction Steps

trying to deploy lambda from asset folder since update to aws-cdk v2

### What did you expect to happen?

deployment properly completed without error

### What actually happened?

an error is raised mentioning that the bundled zip is empty.

### CDK CLI Version

2.1.0

### Framework Version

NA

### Node.js Version

16.13.1

### OS

MacOSx

### Language

Typescript, Go

### Language Version

TS 4.5.2 | Go 1.17.3

**Comment by peterwoodworth** [comment]
Hey, some more context behind what constructs you're using, and reproduction code would be helpful for us to figure out what's going wrong here 😃

**Comment by SamuelCabralCruz** [comment]
I am essentially using the following code to create a lambda function from local directory
```
  new cdk.aws_lambda.Function(scope, constructId.getResourceId(ConstructResourceType.FUNCTION), {
    code: cdk.aws_lambda.Code.fromAsset(bundlingPath, {
      bundling: {
        image: cdk.aws_lambda.Runtime.GO_1_X.bundlingImage,
        user: 'root',
        environment: {
          CGO_ENABLED: '0',
          GOOS: 'linux',
          GOARCH: 'amd64',
        },
        command: ['bash', '-c', [`cd ${bundlingDirectory}`, 'make bundle-cdk'].join(' && ')],
      },
    }),
    handler,
    runtime: cdk.aws_lambda.Runtime.GO_1_X,
  })
```

I forgot to mention that it is not the first time I am experiencing this problem.
Essentially the same problem as: [12667](https://github.com/aws/aws-cdk/issues/12667) and [12536](https://github.com/aws/aws-cdk/issues/12536)

**Comment by SamuelCabralCruz** [comment]
Also, I don't need if it helps, but I found this issue after migrating from @aws-cdk v1.184 to aws-cdk-lib v2.1.0

**Comment by peterwoodworth** [comment]
I'm not able to reproduce this on v2.1.0 - Just to clarify, are you saying that you were not running into this error on `1.134.0`, then you migrated to v2, then this error started occurring without any changes to the lambda?

**Comment by SamuelCabralCruz** [comment]
@peterwoodworth yes, exactly. One important point to mention here, the problem is specific to MacOS environment. Everything is running as expected on my CI/CD pipeline which use a Linux.

**Comment by kaizen3031593** [comment]
@SamuelCabralCruz can you show us your build script? It's a little difficult to diagnose whats going on with just the lambda function, but it is possible that you are not copying the output into the correct location. When you bundle a go lambda you need to make sure the binary gets placed in the `outputDir` like [this](https://github.com/aws-samples/cdk-build-bundle-deploy-example/blob/0b0535d21425a39c31aac5de59570775108579a8/cdk-bundle-go-lambda-example/lib/api-stack.ts#L67).

**Comment by SamuelCabralCruz** [comment]
@kaizen3031593 sure, here is the build script I am using for all my lambdas.
```
.PHONY: bundle-cdk-default
bundle-cdk-default:
	@go mod vendor
	@GOOS=linux go build -mod=vendor -o /asset-output/main
```

This is the "default" implementation for all my makefiles, but it is never overridden.

**Comment by SamuelCabralCruz** [comment]
@kaizen3031593 Is there an important difference between /main and /bootstrap folders?

**Comment by corymhall** [comment]
@SamuelCabralCruz the name of the output just has to match the value you provide in `handler`. The example is using the name `bootstrap` because that is required if you are using the `provided` runtime.

**Comment by SamuelCabralCruz** [comment]
in my case, I am passing in "main" as default argument

**Comment by corymhall** [comment]
@SamuelCabralCruz this could be an issue with caching. Can you look in your `cdk.out` directory for an asset folder with a name something like `asset.xxxxxxsomerealylonghash/`. Inside this folder you should see the `main` binary that was produced by the bundling.

If the asset folder is empty try deleting your `cdk.out` directory and running `cdk synth` again to see if you then see the `main` binary. By default CDK calculates the asset hash based on the source directory and if nothing changes it will skip bundling. If somehow the local asset output got emptied it might cause the issue you are seeing.

If it is not empty try changing something in the go code (whitespace, comment, etc) and try to deploy again. If you deployed an empty zip and then try and deploy a non-empty zip with the same hash, the CDK asset publishing will see that an asset already exists and will skip publishing the new asset.

Alternatively you can change the `assetHashType` to `OUTPUT` in the bundling config which will tell the CDK to calculate the asset hash based on the output binary (which should be different). 

```ts
...
      bundling: {
        assetHashType: cdk.AssetHashType.OUTPUT,
        image: cdk.aws_lambda.Runtime.GO_1_X.bundlingImage,
        user: 'root',
        environment: {
          CGO_ENABLED: '0',
          GOOS: 'linux',
          GOARCH: 'amd64',
        },
        command: ['bash', '-c', [`cd ${bundlingDirectory}`, 'make bundle-cdk'].join(' && ')],
      },
...
```

**Comment by SamuelCabralCruz** [comment]
@corymhall Ok, so I just tried out what you said.
First, I made sure to clean up all vendor directories from my lambdas and I removed the cdk.out folder.
Then, I ran `cdk synth`.
The lambda seems to be bundled properly (Bundling log + go: dependency downloads log), but the command never returned even after long wait.
The cdk.out folder has been created but the asset folder within it was empty.
I also observed that the vendor folder was created in the lambda directory, but was empty.
I tried to run `go mod vendor` in that lambda directory and the vendor folder was properly created with all its content. 
When I rerun the `cdk synth` command then the template is properly outputted.

In short, it seems to indicate that aws-cdk-lib v2.1.0 has some issue with getting rights to write file on disk in remote vendor folder.

I don't think that changing the asset hash type would help here, but still is interesting option I was not aware of.

Hope it helps!

**Comment by sdpoueme** [comment]
I am seeing the same issue  with  cdk version 2.20.0 (build 738ef49).  Below is the code I am building: 
```
import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import * as apigateway from "aws-cdk-lib/aws-apigateway";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as s3 from "aws-cdk-lib/aws-s3";

export class DevopsSspSelfServiceStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const bucket = new s3.Bucket(this, "selfServiceCode");
    const path = require("path");
    
    
    const handler = new lambda.Function(this, "OnnboardingHandler", {
      runtime: lambda.Runtime.NODEJS_14_X, // So we can use async in onboarding.js
      code: lambda.Code.fromAsset(path.join(__dirname, "../resources")),
      handler: "onboarding.main",
      environment: {
        BUCKET: bucket.bucketName,
      },
    });

    bucket.grantReadWrite(handler); // was: handler.role);

    const api = new apigateway.RestApi(this, "onboarding-api", {
      restApiName: "DevOps Self-service",
      description: "Self-service API for DevOps SSP",
    });

    const postOnboardingIntegration = new apigateway.LambdaIntegration(
      handler,
      {
        requestTemplates: { "application/json": '{ "statusCode": "200" }' },
      }
    );

    api.root.addMethod("POST", postOnboardingIntegration); // GET /
  }
}

```

The lambda code is in a folder accessible from the current path.

**Comment by madeline-k** [comment]
Hey @SamuelCabralCruz, are you still seeing this issue? Please comment if so, otherwise this will be automatically closed. Even after closing, if anyone encounters this, please open a new issue.
